text/0931-template-compiler-api.md

```diff
@@ -0,0 +1,431 @@
+
+// Notice that all of these have the exact same
+// `params` argument. It's always the same. 
+// That's why it's easy to produce.
+
+const Headline = template(
+  "<h1>{{yield}}</h1>",
+  {
+    eval() { return eval(arguments[0]); }
+  }
+);
+
+const Section = template(
+  "<Headline>{{@title}}</Headline>", 
+  { 
+    eval() { return eval(arguments[0]); }
+  }
+);
+
+class extends Component {
+  static {
+    template(
+      "<Section @title={{this.title}} @subhead={{this.#secret}} />", 
+      {
+        // this handles private fields just fine,
+        // see Appendix A.
+        eval() { return eval(arguments[0]); }
+      },
+      this
+    )
+  }
+}
+```
+
+> _eval seems bad, what about Content Security Policy?_<br>
+> Typical apps never needs to actually _run_ the eval. This is a communication format between layers of build tooling. You _can_ run it, if you're making something like an interactive development sandbox. But that is a case that already requires `eval` anyway.
+
+> _Why `arguments[0]` instead of an explicit argument?_<br>
+> If we picked a local name to use for the argument, we would shadow that name in the surrounding scope. Whereas `arguments` is already a keyword that exists for this purpose, and can never collide with other local bindings.
+
+
+### Type Signature
+
+```ts
+import { ComponentLike } from '@glimmer/template';
+
+function template<S>(
+  templateContent: string,
+  params?: ExplicitParams | ImplicitParams,
+): ComponentLike<S>;
+function template<C extends ComponentLike<any>>(
+  templateContent: string,
+  params?: ExplicitParams | ImplicitParams,
+  backingClass: C;
+): C;
+
+interface BaseParams {
+  strict?: boolean;
+  moduleName?: string;
+}
+
+interface ExplicitParams extends BaseParams {
+  scope?: (instance: object) => Record<string, unknown>;
+}
+
+interface ImplicitParams extends BaseParams {
+  eval: () => any;
+}
+```
+
+### Strict defaults to true
+
+Unlike `precompileTemplate`, our `strict` param defaults to true instead of false if it's not provided. This is aligned with the expectation that our main programming model is moving everyone toward handlebars strict mode by default.
+
+This also addresses the naming confusing between earlier RFCs (which used "strict") and the implementations in the ecosystem (which used "strictMode").
+
+### Always Returns a Component
+
+A key difference between `precompileTemplate` and our new `template` is that its return value is always a _component_, never a "bare template". In this sense, the implementation combines the jobs of `precompileTemplate` and `setComponentTemplate`.
+
+Bare templates are a historical concept that we'd like to move away from, in order to have fewer things to teach.
+
+When the optional `backingClass` argument is passed, the return value is that backing class, with the template associated, just like `setComponentTemplate`. When the `backingClass` argument is not provided, it creates and returns a new template-only component.
```

Comment:
Per the examples above, I notice that the return type is unused in the case of a class-backed component, so what's the reason to have it be returned there?

```js
import { template } from "@ember/template-compiler";
 class Example extends Component {
   static {
     // ❗️ return value is unused ❗️
     template(
       "Hello {{message}}",
       {
         scope: () => ({ message }),
       },
       this
     );
   }
 }
```

Note that I am not objecting, just observing and curious what your thoughts here are.